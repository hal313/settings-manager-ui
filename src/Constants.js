const ATTRIBUTE_PREFIX = 'data-setting';

const Constants = {
    ATTRIBUTE_PREFIX: ATTRIBUTE_PREFIX,
    ATTRIBUTE_NAME: `${ATTRIBUTE_PREFIX}-name`,
    ATTRIBUTE_TYPE: `${ATTRIBUTE_PREFIX}-type`,
    ATTRIBUTE_CONTAINER_ELEMENT: `${ATTRIBUTE_PREFIX}-container`
};

export {Constants};