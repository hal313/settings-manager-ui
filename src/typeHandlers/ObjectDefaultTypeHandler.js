import { isDefined } from '../util/isDefined.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';
import { getChildSettingElements } from '../util/getChildSettingElements.js';
import { getNameFromElement } from '../util/getNameFromElement.js';

// TODO: Template for adding new child nodes (use a template, if present, for all? (${name}, ${type}, [${value}]))
export class ObjectDefaultTypeHandler {

    constructor() {}

    getType() {
        return 'object:default';
    }

    getValue(element, settingModifier) {
        let childElements = getChildSettingElements(element);
        let valueObject = {};

        // For each child setting, add the name and value touple to the returned settings object
        childElements.forEach((childElement) => {
            // Get the name from the element
            let name = getNameFromElement(childElement);
            // Get the value (which may be a recursive call)
            let value = settingModifier.getValue(childElement);

            // Add the name and value to the value object
            valueObject[name] = value;
        });

        return valueObject;
    }

    setValue(element, settings, settingModifier) {
        // Append the name to the root path
        let childElements = getChildSettingElements(element);
        let childElementNames = childElements.map(getNameFromElement);

        // Remove values from the DOM which do not exist in the value
        childElements.forEach((childElement) => {
            let name = getNameFromElement(childElement);

            // If the DOM has a value which is not in the object,
            if (isDefined(settings)) {
                if (!Object.keys(settings).includes(name)) {
                    childElement.remove();
                }
            }
        });


        // Update values from the values which exist in the DOM
        childElements.forEach((childElement) => {
            let name = getNameFromElement(childElement);

            // If the DOM has a value which is in the settings object
            if (childElementNames.includes(name)) {
                if (isDefined(settings)) {
                    settingModifier.setValue(childElement, settings[name]);
                }
            }
        });


        // Add new child nodes for values which do not exist in the DOM
        if (isDefined(settings)) {
            Object.keys(settings).forEach((name) => {
                let settingValue = settings[name];

                if (!childElementNames.includes(name)) {
                    // Create the element
                    let childElement = settingModifier.createElement(name, settingValue);

                    // Add to the element
                    element.append(childElement);
                }

            });
        }

        element.value = settings;
    }

    createElement(name, value, settingModifier) {
        if (!isDefined(name)) {
            throw createError('The "name" parameter must be specified');
        }

        // Create an element
        const element = document.createElement('div');

        // Set attributes
        element.setAttribute('name', name);
        // Set custom attributes
        element.setAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT, '');

        // Set the value
        this.setValue(element, value, settingModifier);

        return element;
    }

};