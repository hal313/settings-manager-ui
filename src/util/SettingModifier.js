import { getDeclaredTypeFromElement } from './getDeclaredTypeFromElement.js';
import { Constants } from '../Constants.js';
import { StringTextTypeHandler } from '../typeHandlers/StringTextTypeHandler.js';
import { BooleanCheckboxTypeHandler } from '../typeHandlers/BooleanCheckboxTypeHandler.js';
import { NumberNumberTypeHandler } from '../typeHandlers/NumberNumberTypeHandler.js';
import { ObjectDefaultTypeHandler } from '../typeHandlers/ObjectDefaultTypeHandler.js';
import { UndefinedTypeHandler } from '../typeHandlers/UndefinedTypeHandler.js';
import { NullTypeHandler } from '../typeHandlers/NullTypeHandler.js';
import { CollectionObjectTypeHandler } from '../typeHandlers/CollectionObjectTypeHandler.js';
import { StringPasswordTypeHandler } from '../typeHandlers/StringPasswordTypeHandler.js';

// TODO: Tests
export class SettingModifier {

    constructor(typeHandlerManager, typeDecoratorManager) {
        this.typeHandlerManager = typeHandlerManager;
        this.typeDecoratorManager = typeDecoratorManager;


        const stringTextTypeHandler = new StringTextTypeHandler();
        const booleanCheckboxTypeHandler = new BooleanCheckboxTypeHandler();
        const numberNumberTypeHandler = new NumberNumberTypeHandler();
        const objectDefaultTypeHandler = new ObjectDefaultTypeHandler();
        const undefinedTypeHandler = new UndefinedTypeHandler();
        const nullTypeHandler = new NullTypeHandler();
        const collectionObjectTypeHandler = new CollectionObjectTypeHandler();
        const stringPasswordTypeHandler = new StringPasswordTypeHandler();


        this.typeHandlerManager.addTypeHandler(booleanCheckboxTypeHandler);
        this.typeHandlerManager.addTypeHandler(stringTextTypeHandler);
        this.typeHandlerManager.addTypeHandler(numberNumberTypeHandler);
        this.typeHandlerManager.addTypeHandler(objectDefaultTypeHandler);
        this.typeHandlerManager.addTypeHandler(undefinedTypeHandler);
        this.typeHandlerManager.addTypeHandler(nullTypeHandler);
        this.typeHandlerManager.addTypeHandler(collectionObjectTypeHandler);
        this.typeHandlerManager.addTypeHandler(stringPasswordTypeHandler);


        // Add default handlers
        this.typeHandlerManager.setDefaultHandler('string', stringTextTypeHandler.getType());
        this.typeHandlerManager.setDefaultHandler('boolean', booleanCheckboxTypeHandler.getType());
        this.typeHandlerManager.setDefaultHandler('number', numberNumberTypeHandler.getType());
        this.typeHandlerManager.setDefaultHandler('object', objectDefaultTypeHandler.getType());
        this.typeHandlerManager.setDefaultHandler('collection', collectionObjectTypeHandler.getType());

        // TODO: Extract this as a class?
        this.typeDecoratorManager.addTypeDecorators('.*', (element, type, name, value) => {
            element.setAttribute(Constants.ATTRIBUTE_NAME, name);
            element.setAttribute(Constants.ATTRIBUTE_TYPE, type);
        });

        this.path=[];
    }

    getValue(element) {
        let type = getDeclaredTypeFromElement(element);
        let handler = this.typeHandlerManager.getTypeHandler(type);
        return handler.getValue(element, this);
    };

    setValue(element, value) {
        let type = getDeclaredTypeFromElement(element);
        let handler = this.typeHandlerManager.getTypeHandler(type);
        return handler.setValue(element, value, this);
    };

    createElement(name, value) {
        // Get the handler
        let handler = this.typeHandlerManager.inferTypeHandler(name, value);

        // Push the name onto the path
        this.path.push(name);
        // Create the element
        let element = handler.createElement(name, value, this);
        // Set the flatname
        element.setAttribute(Constants.ATTRIBUTE_FLATNAME, this.path.join('.'));
        // Pop the path
        this.path.pop();

        // Apply decorators
        element = this.typeDecoratorManager.applyTypeDecorators(handler.getType(), element, name, value);

        return element;
    };

};
