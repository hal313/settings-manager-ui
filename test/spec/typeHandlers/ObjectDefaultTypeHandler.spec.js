import { ObjectDefaultTypeHandler } from '../../../src/typeHandlers/ObjectDefaultTypeHandler.js';
import { SettingModifier } from '../../../src/util/SettingModifier.js';
import { Constants } from '../../../src/Constants.js';

describe('ObjectDefaultTypeHandler', () => {

    describe('Lifecycle', () => {

        test('should exist', () => {
            expect(ObjectDefaultTypeHandler).toBeDefined();
        });

    });

    describe('API', () => {
        let objectDefaultTypeHandler;
        let settingModifier;

        beforeEach(() => {
            settingModifier = new SettingModifier();
            objectDefaultTypeHandler = settingModifier.typeHandlerManager.getTypeHandlerByType(ObjectDefaultTypeHandler.TYPE);
        });


        describe('getType', () => {

            test('the type is "object:default"', () => {
                expect(objectDefaultTypeHandler.getType()).toEqual(ObjectDefaultTypeHandler.TYPE);
            });

        });

        describe('createElement', () => {

            test('should throw when name is undefined', () => {
                expect(() => objectDefaultTypeHandler.createElement(undefined, 'someValue')).toThrowError();
            });

            test('should throw when name is null', () => {
                expect(() => objectDefaultTypeHandler.createElement(null, 'someValue')).toThrowError();
            });

            test('should throw when name is the empty string', () => {
                expect(() => objectDefaultTypeHandler.createElement('', 'someValue')).toThrowError();
            });

            test('should throw when name is whitespace', () => {
                expect(() => objectDefaultTypeHandler.createElement('   ', 'someValue')).toThrowError();
            });

            test('should create an element with child elements representing the members as input elements', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Check the type
                expect(element instanceof Element).toBeTruthy();
                // Check the required attributes
                expect(element.getAttribute('name')).toEqual(name);
                expect(element.hasAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT)).toBe(true);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should create an element with deeply nested child elements representing the members as input elements', () => {
                const name = 'person';
                const value = {
                    name: {
                        first: 'firstname',
                        last: 'lastname',
                        optional: {
                            middle: 'middlename'
                        }
                    },
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Check the type
                expect(element instanceof Element).toBeTruthy();
                // Check the required attributes
                expect(element.hasAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT)).toBe(true);
                // Check the value
                expect(element.value).toBe(value);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

        describe('getValue', () => {

            test('should get the value from a self-created element', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Check the value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
            });

            test('should get the value from a self-created nested element', () => {
                const name = 'person';
                const value = {
                    name: {
                        first: 'person',
                        last: 'name'
                    },
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Check the value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
            });

            test('should get the value from a self-created deeply nested element', () => {
                const name = 'person';
                const value = {
                    name: {
                        first: 'person',
                        last: 'name',
                        optional: {
                            middle: 'middle'
                        }
                    },
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Check the value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
            });

        });

        describe('setValue', () => {

            test('should set the value for a self-created simple element', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                // Set the new value
                //
                const newValue = JSON.parse(JSON.stringify(value));
                newValue.name = 'name, person';
                delete newValue.age;
                // Set the value
                objectDefaultTypeHandler.setValue(element, newValue, settingModifier);


                // Check the new value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Check the value
                expect(element.value).toBe(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should remove a child element from the DOM when the value no longer exists', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                // Set the new value
                //
                // Create the value
                const newValue = JSON.parse(JSON.stringify(value));
                delete newValue.age;
                // Set the new value
                objectDefaultTypeHandler.setValue(element, newValue, settingModifier);


                // Check the new value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should update a child element value when the value exists in the DOM and in the value object', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();


                // Set the new value
                //
                // Create the new value
                const newValue = JSON.parse(JSON.stringify(value));
                newValue.age = 22;
                newValue.name = 'name, person';
                // Set the value
                objectDefaultTypeHandler.setValue(element, newValue, settingModifier);


                // Check the new value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should add a child element value when the setting does not exist in the DOM and the setting is in the value object', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Validate that the element has the correct value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();


                // Set the new value
                // Create a clone of the old value
                const newValue = JSON.parse(JSON.stringify(value));
                newValue.eyeColor = 'brown';
                // Set the value
                objectDefaultTypeHandler.setValue(element, newValue, settingModifier);

                // Check the new settings object
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should add a child element value when the setting does not exist in the DOM and the setting is in the value object during an update', () => {
                const name = 'person';
                const value = {
                    name: 'person name',
                    age: 21,
                    registeredVoter: true
                };

                // Create the element
                const element = objectDefaultTypeHandler.createElement(name, value, settingModifier);

                // Validate that the element has the correct value
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();


                // Set the new value
                //
                // Create a clone of the old value
                const newValue = JSON.parse(JSON.stringify(value));
                newValue.height = {
                    feet: 5,
                    inches: 10
                };
                // Set the value
                objectDefaultTypeHandler.setValue(element, newValue, settingModifier);


                // Check the new settings object
                expect(objectDefaultTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

    });

});
