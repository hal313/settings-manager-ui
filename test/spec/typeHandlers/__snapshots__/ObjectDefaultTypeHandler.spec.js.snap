// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectDefaultTypeHandler API createElement should create an element with child elements representing the members as input elements 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API createElement should create an element with deeply nested child elements representing the members as input elements 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <div name="name"
       data-setting-name="name"
       data-setting-type="object:default"
       data-setting-container
       class
  >
    <input name="first"
           type="text"
           value="firstname"
           data-setting-name="first"
           data-setting-type="string:text"
    >
    <input name="last"
           type="text"
           value="lastname"
           data-setting-name="last"
           data-setting-type="string:text"
    >
    <div name="optional"
         data-setting-name="optional"
         data-setting-type="object:default"
         data-setting-container
         class
    >
      <input name="middle"
             type="text"
             value="middlename"
             data-setting-name="middle"
             data-setting-type="string:text"
      >
    </div>
  </div>
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should add a child element value when the setting does not exist in the DOM and the setting is in the value object 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should add a child element value when the setting does not exist in the DOM and the setting is in the value object 2`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
  <input name="eyeColor"
         type="text"
         value="brown"
         data-setting-name="eyeColor"
         data-setting-type="string:text"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should add a child element value when the setting does not exist in the DOM and the setting is in the value object during an update 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should add a child element value when the setting does not exist in the DOM and the setting is in the value object during an update 2`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
  <div name="height"
       data-setting-name="height"
       data-setting-type="object:default"
       data-setting-container
       class
  >
    <input name="feet"
           type="number"
           value="5"
           data-setting-name="feet"
           data-setting-type="number:number"
    >
    <input name="inches"
           type="number"
           value="10"
           data-setting-name="inches"
           data-setting-type="number:number"
    >
  </div>
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should remove a child element from the DOM when the value no longer exists 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should remove a child element from the DOM when the value no longer exists 2`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should set the value for a self-created simple element 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should set the value for a self-created simple element 2`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should update a child element value when the value exists in the DOM and in the value object 1`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;

exports[`ObjectDefaultTypeHandler API setValue should update a child element value when the value exists in the DOM and in the value object 2`] = `

<div name="person"
     data-setting-name="person"
     data-setting-type="object:default"
     data-setting-container
     class
>
  <input name="name"
         type="text"
         value="person name"
         data-setting-name="name"
         data-setting-type="string:text"
  >
  <input name="age"
         type="number"
         value="21"
         data-setting-name="age"
         data-setting-type="number:number"
  >
  <input type="checkbox"
         name="registeredVoter"
         data-setting-name="registeredVoter"
         data-setting-type="boolean:checkbox"
         value="true"
  >
</div>

`;
